bot.js

```diff
@@ -23,39 +23,44 @@ logger.add(logger.transports.Console, {
 logger.level = 'debug';
 
 // Add message handling for the bot:
-bot.on('message', function (discordName, discordID, channelID, message, evt) {
-    // Our bot needs to know if it will execute a command
-    // It will listen for messages that will start with `!`
-    if (message.substring(0, 1) === '!') {
-        var args = message.substring(1).split(' ');
+bot.on('message', function (message) {
+    if (message.author.bot) return;
+
+    var discordName = message.member.displayName;
+    var discordID = message.member.id.toString();
+
+    if (message.content.substring(0, 1) === '!') {
+        var args = message.content.substring(1).split(' ');
         var cmd = args[0];
         var argsLeft = args.slice(1);
+        message.channel.startTyping();
         logger.info("Running command: " + cmd);
         switch (cmd) {
             case 'debug': {
                 logger.debug("Discord Name=" + discordName);
                 logger.debug("Discord Id=" + discordID);
-                logger.debug("Channel ID=" + channelID);
-                logger.debug("Message=" + JSON.stringify(message));
+                logger.debug("Channel ID=" + message.channel.id);
+                logger.debug("Message=" + message.content);
                 break;
             }
             case 'update': {
-                updateCommand.run(discordName, discordID, channelID, message, evt, argsLeft);
+                updateCommand.run(discordName, discordID, message, argsLeft);
                 break;
             }
             case 'rank': {
-                rankCommand.run(discordName, discordID, channelID, message, evt, argsLeft);
+                rankCommand.run(discordName, discordID, message, argsLeft);
                 break;
             }
             case 'register': {
-                registerCommand.run(discordName, discordID, channelID, message, evt, argsLeft);
+                registerCommand.run(discordName, discordID, message, argsLeft);
                 break;
             }
             case 'ladder': {
-                ladderCommand.run(discordName, discordID, channelID, message, evt, argsLeft);
+                ladderCommand.run(discordName, discordID, message, argsLeft);
                 break;
             }
             // Just add any case commands if you want to..
         }
+        message.channel.stopTyping();
```

Comment:
I'm not sure this works - it'll stopTyping synchronously after calling run.
A lot of the run() methods though chain async futures / callbacks, ideally we'd call stopTyping once the last step of the run() is complete.

I'm not sure the best way to do that - either just call stopTyping() from within each command's run() once they're done; alternatively, pass in an onComplete function into each command, and get them to call it once the command is done. Then, we can pass in () -> message.channel.stopTyping() as the onComplete function to each.

Reply:
That sounds reasonable. I will change the run functions into promises, which will give us an onComplete and onError function